windows-gaming-driver/src/setup/wizard.rs

```diff
@@ -1,307 +1,34 @@
-                            // bail out if there's already vfio stuff in the file
-                            // or if there's two MODULES lines for some reason
-                            return Ok(false);
-                        } else {
-                            mkic_conf.push(MODULES_PREFIX.to_owned() + KERNEL_MODULES + " " + &line[MODULES_PREFIX.len()..]);
-                            hooks_added = true;
-                        }
-                    } else {
-                        if line.starts_with(HOOKS_PREFIX) && line.contains("modconf") {
-                            *has_modconf = true;
-                        }
-                        mkic_conf.push(line);
-                    }
-                }
-
-                if already_added {
-                    return Ok(true);
-                }
-
-                if hooks_added {
-                    return sudo_write_file(MKINITCPIO_CONF, |writer| {
-                        for line in mkic_conf {
-                            writeln!(writer, "{}", line)?;
-                        }
-                        Ok(())
-                    });
-                }
-            }
-        }
-
-        Ok(false)
-    }
-
-    fn write_vfio_modconf(&self, setup: &SetupConfig) {
-        let vfio_params = setup.vfio_devs.iter().fold(String::new(),
-                                                      |s, i| s + &format!("{:04x}:{:04x},", i.vendor, i.product));
-        assert!(sudo_write_file("/etc/modprobe.d/vfio.conf", |x| {
-            writeln!(x, "options vfio-pci ids={}", vfio_params)
-        }).unwrap_or(false), "Failed to write modconf");
-    }
-
-    fn run(&mut self, cfg: Option<Config>, target: &Path, workdir: &Path, datadir: &Path) {
-        const TROUBLESHOOTING: &'static str = "Troubleshooting (since you apparently already did this):";
-
-        let mut machine = MachineConfig {
-            memory: "".to_owned(),
-            cores: 0,
-            vfio_slots: Vec::new(),
-            network: None,
-            storage: Vec::new(),
-            usb_devices: Vec::new(),
-
-            hugepages: None,
-            threads: None,
-        };
-
-        let mut setup = SetupConfig {
-            cdrom: None,
-            floppy: None,
-            gui: false,
+use setup::iommu;
+use setup::usb;
+use setup::vfio;
+use setup::initramfs;
+use setup::vm;
+use pci_device::PciDevice;
 
-            iommu_commanded: false,
-            reboot_commanded: false,
-            vfio_devs: Vec::new(),
-        };
+struct Wizard;
 
-        if let Some(cfg) = cfg {
-            machine = cfg.machine;
-            if let Some(cfg_setup) = cfg.setup {
-                setup = cfg_setup;
-            }
+impl Wizard {
+    fn run(&mut self, cfg: Option<Config>, cfg_path: &Path, workdir: &Path, datadir: &Path) {
+        let mut config = cfg.unwrap_or_default();
+        if config.setup.is_none() {
+            config.setup = Some(SetupConfig::default());
```

Comment:
As I refactored the `run` method to be quite minimal now and basically only a grouping of functions (which may change in the future), I thought it'd be easier to read the saving of the config and see the config as such.
Additionally, it saves quite a few function calls (which would definitely not be inlined or so) and clones (you know, the CPU is quite slow when it comes to memcopy) :stuck_out_tongue: